### PR TITLE
Native Types: add support for non integer arguments

### PR DESCRIPTION
**Pull request by carmenberndt**
closes https://github.com/prisma/prisma-engines/issues/1145

### CODE REVIEW

File: libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs [hunk 1]
```diff
@@ -164,7 +166,13 @@ impl Connector for MySqlDatamodelConnector {
         &self.constructors
     }
 
-    fn parse_native_type(&self, name: &str, args: Vec<u32>) -> Result<NativeTypeInstance, ConnectorError> {
+    fn parse_native_type(&self, name: &str, args: Vec<String>) -> Result<NativeTypeInstance, ConnectorError> {
+        let res = parse_u32_arguments(args);
+        if res.is_err() {
+            return Err(res.err().unwrap());
+        }
+        let parsed_arg = res.unwrap();
```
Comment (pimeys) [inline libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs:174]: Could this work better like this?

```rust
let parsed_arg = parse_u32_arguements(args)?;
```

If not, why?

File: libs/datamodel/connectors/sql-datamodel-connector/src/postgres_datamodel_connector.rs [hunk 2]
```diff
@@ -132,30 +133,36 @@ impl Connector for PostgresDatamodelConnector {
         &self.constructors
     }
 
-    fn parse_native_type(&self, name: &str, args: Vec<u32>) -> Result<NativeTypeInstance, ConnectorError> {
+    fn parse_native_type(&self, name: &str, args: Vec<String>) -> Result<NativeTypeInstance, ConnectorError> {
+        let res = parse_u32_arguments(args);
+        if res.is_err() {
+            return Err(res.err().unwrap());
+        }
+        let parsed_arg = res.unwrap();
```
Comment (pimeys) [inline libs/datamodel/connectors/sql-datamodel-connector/src/postgres_datamodel_connector.rs:141]: Could this work better like this?

```rust
let parsed_arg = parse_u32_arguements(args)?;
```

If not, why?

**Review by pimeys — CHANGES_REQUESTED** [review]

File: libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs (hunk 1 above)
Comment (carmenberndt) [reply libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs]: yes, that's a lot better!

File: libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs [hunk 3]
```diff
@@ -178,45 +182,45 @@ impl Connector for MySqlDatamodelConnector {
             BIG_INT_TYPE_NAME => MySqlType::BigInt,
             UNSIGNED_BIG_INT_TYPE_NAME => MySqlType::UnsignedBigInt,
             DECIMAL_TYPE_NAME => {
-                if let (Some(first_arg), Some(second_arg)) = (args.get(0), args.get(1)) {
```
Comment (pimeys) [inline libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs:181]: Does this here mean you can have either two or more parameters without a validation error? At least it looks like that! Also, I'd like to point you to a concept of slice patterns. They look like to be a nicer solution here as least how it looks, but if I'm right about the bug of allowing more than two params here (and the matches afterwards), we'll fix two problems here! See:

```rust
DECIMAL_TYPE_NAME => match parsed_args.as_slice() {
    // Matches two and only two params.
    [scale, precision] => MySqlType::Decimal(*scale, *precision),
    // Matches all other parameter numbers.
    _ => {
        return Err(ConnectorError::new_argument_count_mismatch_error(
            DECIMAL_TYPE_NAME,
            2,
            parsed_args.len(),
        ));
    }
},
```

**Review by pimeys — CHANGES_REQUESTED** [review]
Another comment, this takes a bit more to fix. Meanwhile I'm going for one other problem I found out!

File: libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs (hunk 3 above)
Comment (pimeys) [reply libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs:181]: https://doc.rust-lang.org/reference/patterns.html#slice-patterns

Comment (carmenberndt) [reply libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs:181]: No, there is a validation before, checking the number of arguments. So the `argument_count_mismatch_error` will never be called actually, I guess this should be removed here? It felt wrong to not error here though 🤷 

(here's the validation before: https://github.com/prisma/prisma-engines/blob/d6e1f3b44115a01f67b4f9cc7123783065db4837/libs/datamodel/core/src/transform/ast_to_dml/lift.rs#L295)

Apart from that, the slice pattern does look cleaner, I'll use that.

File: libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs [hunk 4]
```diff
@@ -257,9 +261,9 @@ impl Connector for MySqlDatamodelConnector {
             MEDIUM_TEXT_TYPE_NAME => MySqlType::MediumText,
             LONG_TEXT_TYPE_NAME => MySqlType::LongText,
             DATE_TYPE_NAME => MySqlType::Date,
-            TIME_TYPE_NAME => MySqlType::Time(args.first().cloned()),
-            DATETIME_TYPE_NAME => MySqlType::DateTime(args.first().cloned()),
-            TIMESTAMP_TYPE_NAME => MySqlType::Timestamp(args.first().cloned()),
+            TIME_TYPE_NAME => MySqlType::Time(parsed_args.first().cloned()),
```
Comment (pimeys) [inline libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs:264]: We could also do validation in here, maybe? 

```
            LONG_TEXT_TYPE_NAME => match parsed_args.as_slice() {
                [] => MySqlType::LongText,
                _ => todo!("Maybe return an error?"),
            },
            TIME_TYPE_NAME => match parsed_args.as_slice() {
                [fractions] => MySqlType::Time(Some(fractions)),
                _ => todo!("Return an error"),
            }
```

Comment (pimeys) [reply libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs:264]: I mean, this is already a bigger refactor, but we're matching anyhow here, so would make sense if the param count validation would also be here?

Comment (carmenberndt) [reply libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs:264]: yes, for consistency it would also be good here 👍

**Comment by carmenberndt** [comment]
> How do we handle optional parameters in the errors? Here it looks like we expect 0 params, but also one is allowed...

Good point, maybe it's best to add another error type for native types with optional parameters.

**Pull request merged**